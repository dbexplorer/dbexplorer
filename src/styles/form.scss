.jsde-form {
    padding: 8px;
    display: table;
    border-collapse: separate;
    border-spacing: 8px
}

.jsde-field {
    display: table-row;
}

.jsde-field__label {
    display: table-cell;
    max-width: 200px;
}

.jsde-field__label--ref {
    text-decoration: underline;
    text-decoration-style: dashed;
    cursor: pointer;
}

.jsde-field__input {
    display: table-cell;
}

.jsde-relationship {
    padding: 8px 0;
}