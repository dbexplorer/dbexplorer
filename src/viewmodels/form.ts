import { cssPrefix } from "../utils";
import { IBaseViewModel } from "./base";
import { TableColumnDescription } from "./table";

export class FormRelationshipDescription {
  title: string;
  entity: string;
  key: string | string[];
}
export class FormStringFieldViewModel {
  getName() {
    return this.columnDescription.name;
  }

  getTitle() {
    return this.columnDescription.title;
  }
  getText() {
    return this.text;
  }
  setText(text: string) {
    if (this.text !== text) {
      this.text = text;
      this.updateCallback(this.text);
    }
  }
  constructor(
    private columnDescription: TableColumnDescription,
    private text: string = null
  ) { }
  css() {
    return {
      root: cssPrefix("field"),
      label: cssPrefix("field__label"),
      input: cssPrefix("field__input"),
    };
  }
  updateCallback: (data: string) => any;
}

export class FormRelationshipViewModel {
  getTitle() {
    return this.relDescription.title;
  }
  getEntityId() {
    return this.relDescription.entity;
  }
  getKeyField() {
    return this.relDescription.key;
  }
  constructor(
    private relDescription: FormRelationshipDescription,
  ) { }
  css() {
    return {
      root: cssPrefix("relationship"),
      label: cssPrefix("relationship__label"),
    };
  }
  updateCallback: (data: string) => any;
  exploreCallback: () => void;
}

export class FormViewModel implements IBaseViewModel {
  fields: FormStringFieldViewModel[] = [];
  rels: FormRelationshipViewModel[] = [];
  constructor(private fieldsDescription: TableColumnDescription[], relDescriptions: FormRelationshipDescription[]) {
    this.fields = fieldsDescription.map(
      (field) => new FormStringFieldViewModel(field)
    );
    if (relDescriptions) {
      this.rels = relDescriptions.map(
        (rel) => {
          const relation = new FormRelationshipViewModel(rel)
          relation.exploreCallback = () => {
            this.exploreRelationshipCallback(relation);
          }
          return relation;
        }
      );
    }
  }
  css() {
    return {
      root: cssPrefix("form"),
    };
  }
  reloadData() {
    this.getDataCallback((data: any) => {
      let dataRow = data[0];
      this.fields.forEach((field) =>
        field.setText(dataRow[field.getName()].toString())
      );
    });
  }
  getDataCallback: (ready: any) => void;
  exploreRelationshipCallback: (row: FormRelationshipViewModel) => any;
}
