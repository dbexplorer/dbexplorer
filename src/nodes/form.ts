import { FormStringFieldViewModel, FormViewModel } from "../viewmodels/form";

export class FormInputFieldNode {
  private inputHTMLElement: HTMLInputElement;
  private labelHTMLElement: HTMLLabelElement;
  private fieldHTMLElement: HTMLDivElement;
  constructor(private viewModel: FormStringFieldViewModel) {
    this.labelHTMLElement = document.createElement("label");
    this.inputHTMLElement = document.createElement("input");
    this.fieldHTMLElement = document.createElement("div");
    this.fieldHTMLElement.appendChild(this.labelHTMLElement);
    this.fieldHTMLElement.appendChild(this.inputHTMLElement);

    this.labelHTMLElement.innerHTML = viewModel.getTitle();
    this.inputHTMLElement.innerHTML = viewModel.getText();
    viewModel.updateCallback = (data: string) => {
      this.inputHTMLElement.innerHTML = data;
    };
  }
  element(): HTMLElement {
    return this.fieldHTMLElement;
  }
}

export class FormNode {
  private formHTMLElement: HTMLFormElement;
  private fieldNodes: FormInputFieldNode[] = [];
  constructor(private viewModel: FormViewModel) {
    this.formHTMLElement = document.createElement("form");
    this.formHTMLElement.action = "#";
    viewModel.fields.forEach((field) => {
      let fieldNode = new FormInputFieldNode(field);
      this.fieldNodes.push();
      this.formHTMLElement.appendChild(fieldNode.element());
    });
  }
  element(): HTMLElement {
    return this.formHTMLElement;
  }
}
