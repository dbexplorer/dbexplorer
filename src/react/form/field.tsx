import React from 'react';
import { FormStringFieldViewModel } from '../../viewmodels/form';

interface IProps {
  model: FormStringFieldViewModel;
}

interface IState {
  title: string;
  text: string;
}
export class FormInputField extends React.Component<IProps, IState> {
  constructor(props: any) {
    super(props);
    this.handleChange = this.handleChange.bind(this)
    const model = props.model as FormStringFieldViewModel;
    this.state = { title: model.getTitle(), text: model.getText() };
    model.updateCallback = (data: string) => {
      this.setState({ text: data });
    };
  }
  handleChange(e: any) {
    this.props.model.setText(e.target.value);
  }
  render() {
    return (
      <div>
        <label>{this.state.title}</label>
        <input value={this.state.text} onChange={this.handleChange}></input>
      </div>
    );
  }
}