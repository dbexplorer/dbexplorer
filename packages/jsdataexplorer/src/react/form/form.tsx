import React, { useEffect } from 'react';
import { FormViewModel } from '../../viewmodels/form';
import { FormInputField } from './field';
import { FormRelationship } from './relationship';

export function Form({ model }: { model: FormViewModel }) {
  const css = model.css();
  useEffect(() => {
    model.reloadData();
  }, [])
  return (
    <form className={css.root} action="#">
      {model.fields.map((f, index) => <FormInputField key={index} model={f} />)}
      {model.rels.map((r, index) => <FormRelationship key={index} model={r} />)}
    </form>
  );
}