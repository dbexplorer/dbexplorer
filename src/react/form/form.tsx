import React from 'react';
import { FormViewModel } from '../../viewmodels/form';
import { FormInputField } from './field';

interface IProps {
  model: FormViewModel;
}

export class Form extends React.Component<IProps> {
  constructor(props: any) {
    super(props);
  }
  css() {
    return this.props.model.css();
  }
  render() {
    return (
      <form className={this.css().root} action="#">
        {this.props.model.fields.map((f, index) => <FormInputField key={index} model={f} />)}
      </form>
    );
  }
}