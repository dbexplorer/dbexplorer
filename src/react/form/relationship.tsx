import React from 'react';
import { FormRelationshipViewModel } from '../../viewmodels/form';

interface IProps {
  model: FormRelationshipViewModel;
}

interface IState {
  text: string;
}
export class FormRelationship extends React.Component<IProps, IState> {
  constructor(props: any) {
    super(props);
    this.handleClick = this.handleClick.bind(this)
    this.state = { text: props.model.getTitle() };
  }
  css() {
    return this.props.model.css()
  }
  handleClick() {
    this.props.model.exploreCallback();
  }
  render() {
    return (
      <div className={this.css().root}>
        <button className={this.css().label} onClick={this.handleClick}>{this.props.model.getTitle()}</button>
      </div>
    );
  }
}