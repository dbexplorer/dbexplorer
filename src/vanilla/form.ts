import { FormStringFieldViewModel, FormViewModel } from "../viewmodels/form";

export class FormInputFieldNode {
  private inputHTMLElement: HTMLInputElement;
  private labelHTMLElement: HTMLLabelElement;
  private fieldHTMLElement: HTMLDivElement;
  constructor(private viewModel: FormStringFieldViewModel) {
    this.labelHTMLElement = document.createElement("label");
    this.inputHTMLElement = document.createElement("input");
    this.fieldHTMLElement = document.createElement("div");
    this.fieldHTMLElement.appendChild(this.labelHTMLElement);
    this.fieldHTMLElement.appendChild(this.inputHTMLElement);

    this.labelHTMLElement.innerHTML = viewModel.getTitle();
    this.inputHTMLElement.value = viewModel.getText();
    viewModel.updateCallback = (data: string) => {
      this.inputHTMLElement.value = data;
    };
  }
  element(): HTMLElement {
    return this.fieldHTMLElement;
  }
}

export class FormNode {
  private formHTMLElement: HTMLFormElement;
  private fieldNodes: FormInputFieldNode[] = [];
  constructor(private viewModel: FormViewModel) {
    this.formHTMLElement = document.createElement("form");
    this.formHTMLElement.action = "#";
    viewModel.fields.forEach((field) => {
      let fieldNode = new FormInputFieldNode(field);
      this.fieldNodes.push();
      this.formHTMLElement.appendChild(fieldNode.element());
    });
    viewModel.reloadData();
  }
  element(): HTMLElement {
    return this.formHTMLElement;
  }
}
