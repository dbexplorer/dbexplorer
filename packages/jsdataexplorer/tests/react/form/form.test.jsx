import React from 'react';
import { Form } from "../../../src/react/form/form";
import { render } from '@testing-library/react'
import {
  FormViewModel
} from "../../../src/viewmodels/form";

jest.mock("../../../src/react/form/field", () => ({
  FormInputField: (props) => {
    return <mock-form-input-field title={props.model.getTitle()} />;
  }
}));
jest.mock("../../../src/react/form/relationship", () => ({
  FormRelationship: (props) => {
    return <mock-form-relationship title={props.model.getTitle()} />;
  }
}));

test("Form HTML test", () => {
  var formViewModel = new FormViewModel([

    {
      name: "f1",
      title: "field_1"
    },
    {
      name: "f2",
      title: "field_2"
    },
    {
      name: "f3",
      title: "field_3"
    }
  ], [{
    title: "r 1",
    childTable: "child1",
    childKey: "e_key"
  },
  {
    title: "r 2",
    childTable: "child2",
    childKey: "e_key"
  }]);

  formViewModel.getDataCallback = (ready) => { };

  const { container } = render(<Form model={formViewModel} />);
  container.querySelector("form")
  expect(container.firstChild).toMatchSnapshot();
});
