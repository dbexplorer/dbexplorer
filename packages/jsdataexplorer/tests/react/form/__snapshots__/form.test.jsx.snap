// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form HTML test 1`] = `
<form
  action="#"
  class="jsde-form"
>
  <mock-form-input-field
    title="field_1"
  />
  <mock-form-input-field
    title="field_2"
  />
  <mock-form-input-field
    title="field_3"
  />
  <mock-form-relationship
    title="r 1"
  />
  <mock-form-relationship
    title="r 2"
  />
</form>
`;
